naps/cores/cmv12k/s7_phy.py

```diff
@@ -0,0 +1,471 @@
+        self.lane_pattern = 0xA95
+        # switch (only) sensor sequencer to 12 bit mode
+        sensor.write_reg(118, 0)
+
+    @driver_method
+    def initial_alignment(self):
+        # try to align just the control lane
+        self.set_ctrl_delay(1, 16) # set clock delay to middle position
+
+        for half in range(2):
+            for slip in range(6):
+                self.set_ctrl_delay(0, 0) # zero control lane delay
+                for delay in range(32):
+                    if self.ctrl_lane_match:
+                        return True
+                    self.ctrl_lane_delay_inc = 1 # bump control lane delay
+                self.ctrl_lane_bitslip = 1 # slip control lane by one bit
+            self.ctrl_lane_bitslip = 2 # swap data half words
+
+        return False
+
+    @driver_method
+    def clock_alignment(self):
+        num_lanes = 32
+        self.set_ctrl_delay(1, 0) # zero clock delay
+
+        best_clock_delay = 0
+        best_clock_match = 0
+        # try out all the possible clock delays
+        for clock_delay in range(32):
+            # we want the clock delay for which the lanes match the "most" in
+            # order to give us the widest latitude for adjustment.
+
+            # first, we evaluate each lane by choosing the bitslip value for
+            # which the lane matches over the maximum number of delays
+            best_match_count = [0]*num_lanes
+            for bitslip in range(6):
+                match_count = [0]*num_lanes
+                self.data_lane_delay_reset = (1<<num_lanes)-1 # zero all lane delays
+                for lane_delay in range(32):
+                    all_matches = self.data_lane_match
+                    for lane in range(num_lanes):
+                        match_count[lane] += 1 if all_matches & (1 << lane) else 0
+                    self.data_lane_delay_inc = (1<<num_lanes)-1 # increment all delays
+                for lane in range(num_lanes):
+                    if best_match_count[lane] < match_count[lane]:
+                        best_match_count[lane] = match_count[lane]
+                self.data_lane_bitslip = (1<<num_lanes)-1 # slip all bits
+            # then we measure the "most"ness for this clock delay as which lane matched the fewest times
+            clock_match = min(best_match_count)
+            if best_clock_match < clock_match:
+                best_clock_match = clock_match
+                best_clock_delay = clock_delay
+
+            print("[{:02d}] = {:02d}".format(clock_delay, clock_match))
+            self.ctrl_lane_delay_inc = 2
+
+        self.set_ctrl_delay(1, best_clock_delay)
+        return best_clock_delay, best_clock_match
+
+    @driver_method
+    def data_alignment(self):
+        num_lanes = 32
+
+        # determine the best bitslip as the one which matches over the most delays
+        best_bitslip = [0]*(num_lanes+1)
+        best_bitslip_matches = [0]*(num_lanes+1)
+        for bitslip in range(6):
```

Comment:
This repeats what was done in the `clock_alignment`. Why not just save the best bitslip already in the `clock_alignment` step and simply apply it here?